cmd/car/compile.go

```diff
@@ -0,0 +1,446 @@
+			outFileName = carName
+		}
+
+		if outFileName == "-" && !c.IsSet("output") {
+			return fmt.Errorf("cannot stream carv2's to stdout")
+		}
+		bs, err := blockstore.OpenReadWrite(outFileName, roots)
+		if err != nil {
+			return err
+		}
+		for bc, blk := range outBlocks {
+			ob, _ := blocks.NewBlockWithCid(blk, bc)
+			bs.Put(c.Context, ob)
+		}
+		return bs.Finalize()
+	}
+
+	return nil
+}
+
+func serializeBlock(ctx context.Context, codec cid.Prefix, encoding string, raw []byte) (cid.Cid, []byte, error) {
+	ls := cidlink.DefaultLinkSystem()
+	store := memstore.Store{Bag: map[string][]byte{}}
+	ls.SetReadStorage(&store)
+	ls.SetWriteStorage(&store)
+	b := basicnode.Prototype.Any.NewBuilder()
+	if encoding == "json" {
+		if err := dagjson.Decode(b, bytes.NewBuffer(raw)); err != nil {
+			return cid.Undef, nil, err
+		}
+	} else if encoding == "raw" {
+		if err := b.AssignBytes(raw); err != nil {
+			return cid.Undef, nil, err
+		}
+	} else {
+		return cid.Undef, nil, fmt.Errorf("unknown encoding: %s", encoding)
+	}
+	lnk, err := ls.Store(linking.LinkContext{Ctx: ctx}, cidlink.LinkPrototype{Prefix: codec}, b.Build())
+	if err != nil {
+		return cid.Undef, nil, err
+	}
+	outCid := lnk.(cidlink.Link).Cid
+	outBytes, outErr := store.Get(ctx, outCid.KeyString())
+	return outCid, outBytes, outErr
+}
+
+// DebugCar is a command to translate between a car file, and a human-debuggable patch-like format.
+func DebugCar(c *cli.Context) error {
+	var err error
+	inStream := os.Stdin
+	inFile := "-"
+	if c.Args().Len() >= 1 {
+		inFile = c.Args().First()
+		inStream, err = os.Open(inFile)
+		if err != nil {
+			return err
+		}
+	}
+
+	rd, err := carv2.NewBlockReader(inStream)
+	if err != nil {
+		return err
+	}
+
+	// patch the header.
+	outStream := os.Stdout
+	if c.IsSet("output") {
+		outFileName := c.String("output")
+		outFile, err := os.Create(outFileName)
+		if err != nil {
+			return err
+		}
+		defer outFile.Close()
+		outStream = outFile
+	}
+
+	outStream.WriteString("car compile ")
+	if rd.Version == 2 {
+		outStream.WriteString("--v2 ")
+	}
+
+	outStream.WriteString(inFile + "\n")
+	for _, rt := range rd.Roots {
+		fmt.Fprintf(outStream, "root %s\n", rt.String())
+	}
+
+	// patch each block.
+	nxt, err := rd.Next()
+	if err != nil {
+		return err
+	}
+	for nxt != nil {
+		chunk, err := patch(c.Context, nxt.Cid(), nxt.RawData())
+		if err != nil {
+			return err
+		}
+		outStream.Write(chunk)
+
+		nxt, err = rd.Next()
+		if err == io.EOF {
+			return nil
+		}
+	}
+
+	return nil
+}
+
+func patch(ctx context.Context, c cid.Cid, blk []byte) ([]byte, error) {
+	ls := cidlink.DefaultLinkSystem()
+	store := memstore.Store{Bag: map[string][]byte{}}
+	ls.SetReadStorage(&store)
+	ls.SetWriteStorage(&store)
+	store.Put(ctx, c.KeyString(), blk)
+	node, err := ls.Load(linking.LinkContext{Ctx: ctx}, cidlink.Link{Cid: c}, basicnode.Prototype.Any)
+	if err != nil {
+		return nil, fmt.Errorf("could not load block: %q", err)
+	}
+
+	outMode := "json"
```

Comment:
maybe we should make this `"dag-json"` to be clear about what's going on and give optionality in the future--if we're specific about codec now then we have more flexibility later

while it's true that this is not strict dag-json because it's pretty-printed, it's still dag-json compatible for decoders.